components/gitpod-protocol/go/conn_ws.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) 2020 Jaime Pillora <dev@jpillora.com>. All rights reserved.
+ *  Licensed under the MIT License. See https://github.com/jpillora/chisel/blob/7aa0da95db178b8bc4f20ab49128368348fd4410/LICENSE for license information.
+ *--------------------------------------------------------------------------------------------*/
+// copied and modified from https://github.com/jpillora/chisel/blob/33fa2010abd42ec76ed9011995f5240642b1a3c5/share/cnet/conn_ws.go
+package protocol
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+type WsConn struct {
+	*websocket.Conn
+	buff []byte
+
+	Ctx    context.Context
+	cancel func()
+
+	once     sync.Once
+	closeErr error
+	waitDone chan struct{}
+}
+
+var (
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+
+	// Time allowed to read the next pong message from the peer.
+	pongWait = 15 * time.Second
+
+	// Send pings to peer with this period. Must be less than pongWait.
+	pingPeriod = (pongWait * 9) / 10
+)
+
+//NewWebSocketConn converts a websocket.Conn into a net.Conn
+func NewWebSocketConn(ctx context.Context, websocketConn *websocket.Conn, onStale func(staleErr error)) *WsConn {
+	cCtx, cancel := context.WithCancel(ctx)
+	c := &WsConn{
+		Conn:     websocketConn,
+		waitDone: make(chan struct{}),
+		Ctx:      cCtx,
+		cancel:   cancel,
+	}
+	c.SetReadDeadline(time.Now().Add(pongWait))
```

Comment:
Handle errors from underlying conn?

Reply:
I am not sure it is necessary, in official examples they just call them: https://github.com/gorilla/websocket/blob/e8629af678b7fe13f35dff5e197de93b4148a909/examples/command/main.go#L44-L46 I think underlying implementation does not actually throw but only set values.

Reply:
https://golang.org/src/net/net.go?s=8086:8135#L236

Main concern is that it is critical to the heartbeat but indicative of a deeper issue that would likely throw errors elsewhere as well.